### ISSUE TITLE
C/C++ dependency analysis can handle all of the boost library headers

### ISSUE THREAD

**Issue by adammurdoch** [gradle/gradle-native#301]
Split out of #279, as an additional piece of work. The boost libraries pull some interesting preprocessor tricks that currently baffle the C/C++ dependency analysis, including the improvements implemented as part of #279, so that the analysis gives up and treats all of the include directories as inputs. This kills the developer experience.

The boost libraries are super popular and so we should invest some effort in improving the dependency analysis so that it can understand all of the boost library headers.

**Comment by adammurdoch** [comment]
I spend a bit more time looking into this. I think it would be good to change approach to a greater or lesser degree, and we should return to this after tackling #302 (aka make what we have faster).

Specifically, it seems all of the complexity in the boost headers is actually hidden behind some simple `#ifdef` checks for macros that are only used for development of boost itself. So, rather than trying to understand all of the complex `#define` expressions in boost we might instead understand simple `#ifndef` expressions and ignore all the complexity. We might do this by continuing to understand the source ourselves, asking the compiler to do the preprocessing or both (with one being the fallback for the other, as we can't use one of the approaches for everything).

Regardless of the implementation, this analysis is going to need to know the macros that will be defined at compile time, which are not present on the `Depend` task. I'd defer changing this until we know the `Depend` task's fate, as #302 may see it disappear or reworked in a major way (or not).

**Comment by schwaerz** [comment]
@lacasseio Will this issue be dealt with in #302? I'm getting the MD5 error because the boost headers cannot be parsed and all include directories are considered which in turn would include the .gradle directory...
Not sure which ticket is the one to watch. Can you please clarify? Thanks :)

**Comment by lacasseio** [comment]
Yes, it's a common issue users have being reporting. Could you confirm this is what you are seeing: Basically, you are using some Boost library and including headers that Gradle fails to parse because it finds something like `#include MACRO` and your includes are at the root of your project where the `.gradle` folder get generated.

The workaround would be to ensure your includes aren't in your project directory but at least one folder down (such `includes). It's far from ideal but it's the best solution.

To help direct our work toward a better fix, could you tell us what libraries and possibly `#include` you have inside the Boost library. Better yet, if you could provide a sample project, with just the C/C++/headers and their respective `#include` directives. The code itself (functions, etc.) is relevant for us to reproduce your issue.

**Comment by schwaerz** [comment]
@lacasseio Thanks for the quick response - and sorry for the delay. :flushed: 

Actually as you assumed we're using some boost headers for our application. Currently the folder containing the `.gradle` directory has to be included as some include headers get referenced by `<folder_name/header.h>`. Since `folder_name` is at the same level like `.gradle` we get the issue.

Moving the headers into different folders would probably be a solution. I tried it for one module and it seems to work.

About the example: Not sure what exactly I should provide. Will an application which uses the folder where `.gradle` is located as include directory + uses some header files containing macros in at least one of the `#include` directives be fine? Or does it have to be boost?
Unfortunately I cannot send any code as is so I wanted to create a small example project which reproduces the issue. However while working on that example I realized that I'm not 100% sure about what would help here.

Would a very small example - without boost - help?
